### ISSUE TITLE
[Animal] Maintaining Herd Size by Animal Culling

### ISSUE THREAD

**Issue by allisterakun** [RuminantFarmSystems/RuFaS#1684]
## Current System ##
1. `Cow.cull_update()`
    1. If a cow's estimated daily milk production value falls below a threshold specified by user input `animal.animal_config.management_decisions.cull_milk_production`, the cow will be culled for low milk production.
    2. If the cow's age reaches the future cull date, the cow will be culled.
    3. If the cow's age reaches the future death date, the cow will be culled due to death.
2. `LifeCycleManager._check_if_heifers_need_to_be_sold()`
    1. Keep selling the last heiferIII in the array until the herd size (the total number of heiferIIIs and cows) is no more than 3% over the user-defined herd size or until there are no more heiferIIIs in the herd.
3. `LifeCycleManager._check_if_replacement_heifers_needed()`
    1. Keep buying replacement cows until the herd size exceeds 1% over the user-defined herd size.
## Proposed Changes ##
1. See [Design doc: Update method of maintaining herd size](https://docs.google.com/document/d/1bgwRZTbV7Me6gcLyBxHs06njqheowCgoVKe9J7lEM4Q/edit?usp=sharing) for details
2. Keep the current logic about culling due to cows reaching the future cull date or future death date.
3. [Nice-To-Have] Sell the lowest-performing cows
    1. Have two new user inputs for the tolerance for selling and buying herd size
    2. Create a new function in HerdManager that checks the herd size against the desired number for every user-defined period of time.
        1. If the herd size is too small
            1. Buy more heiferIII until the herd size reaches the ideal number (within the user-defined tolerance).
            2. Implement buying logic:
                1. Purchasing criteria according to body weight, genetics, and production levels
                2. Discuss with the SMEs for details
            3. Change the replacement animals to heiferIIIs (currently, it is cows).
        2. If the herd size is OK
            1. 🙂
        3. If the herd size is too large
            1. Sort all lactating cows by their milk production level.
            2. Sell the bottom least-performing cows to keep the herd size within the user-defined tolerance.

![image](https://github.com/RuminantFarmSystems/MASM/assets/39185942/654e7f78-5e16-41e7-90a3-26333ed0792d)
https://whimsical.com/design-doc-diagram-Vy2gnn6i59fJRGjKKTMnxR 
**Target Deadline**
By when should this issue be done.

**Comment by ew3361zh** [comment]
Note: Still blocked and need more inputs as of 2/19/25.
It would be better to work on this issue after #1680 has been completed.

**Comment by jadamchick** [comment]
Hi, a couple quick thoughts -- 
- is there more background somewhere on the idea of "buying logic" ("Purchasing criteria according to body weight, genetics, and production levels" as it states above)? I don't see this mentioned in the design doc, and it would be helpful to hear more explanation for why to do something different than choosing replacement heifers at random from the pool/market
- we will need to be thoughtful about what metric is used for sorting/ranking cows to determine who gets shipped. Sorting on production makes sense (the M305 discussion [here ](https://github.com/RuminantFarmSystems/RuFaS/issues/2012)is relevant); there will probably need to be some other filters and/or immunity factors based on repro status, etc.?

**Comment by krb78** [comment]
Hi team! A few things
1. when the culling model is FULLY "fixed" we will need to adjust the "parity" culls too because those will also only happen if/when a new animal enters the lactation. perhaps this will need the animal disease model. until then, i think we are ok because the "sick cow" will leave and then a recently calved first lactation animal will fill her spot. only if there are additional first lactation animals do we complete the above logic. 
2. I'm not sure what the "buying logic" is either (if I wrote it, I sincerely apologize for confusing everyone!). one doesn't get to selectively purchase "high production" or "higher genetic merit" animals in our base case (perhaps someday this can be an option, but it certainly is not appropriate today). RuFaS should purchase a heiferIII that is in the typical distribution of heiferIIIs that would be in the herd. In other words, I think we keep the current code / process for purchasing.
